### PR TITLE
Swagger backend REST API

### PR DESCRIPTION
**Pull request by jordigilh**
Swagger definition for the backend REST API defined in https://github.com/project-flotta/flotta-operator/pull/236

### CODE REVIEW

File: swagger-backend.yaml [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        type: boolean
+  device-registration-response:
+    type: object
+    properties:
+      target-namespace:
+        type: string
+      message:
+        type: string
+paths:
+  '/namespaces/{namespace}/devices/{device-id}/deregister':
```
Comment (jakub-dzon) [inline swagger-backend.yaml:42]: I don't think that `namespaces` path segment is needed.

Comment (jakub-dzon) [inline swagger-backend.yaml:42]: `.../deregistrable` or `.../finalize` (if you agree with [my comment](https://github.com/project-flotta/flotta-operator/pull/236/files/d712fed91ac4c26188efc552de3230ba6123160d#r884143325)

Comment (jakub-dzon) [reply swagger-backend.yaml:42]: Or maybe general path pattern like `/devices/{namespace}/{device-id}/...`?

File: swagger-backend.yaml [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+          type: string
+          required: true
+        - in: body
+          name: enrolment-info
+          required: true
+          schema:
+            $ref: 'swagger.yaml#/definitions/enrolment-info'
+      responses:
+        '200':
+          description: Success
```
Comment (jakub-dzon) [inline swagger-backend.yaml:118]: We need to somehow encode in the response that the device is already present - either in the body or status code. Maybe HTTP 201 (Created) when new enrollment is recoded and HTTP 200 (OK) when the device already exists?

File: swagger-backend.yaml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            $ref: '#/definitions/device-registration-response'
+        '401':
+          description: Unauthorized
+        '403':
+          description: Forbidden
+        default:
+          description: Error
+          schema:
+            $ref: '#/definitions/device-registration-response'
+    put:
```
Comment (jakub-dzon) [inline swagger-backend.yaml:160]: :+1:

File: swagger-backend.yaml [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+          description: Namespace where the device resides
+          type: string
+          required: true
+        - in: path
+          name: device-id
+          description: Device ID
+          type: string
+          required: true
+        - in: body
+          name: registration-information
```
Comment (jakub-dzon) [inline swagger-backend.yaml:143]: That should be for `put` and what's in `put` should be here.

Comment (jakub-dzon) [reply swagger-backend.yaml:143]: Scratch that...

**Comment by jakub-dzon** [comment]
The heartbeat operation is missing.

File: swagger-backend.yaml (hunk 1 above)
Comment (jordigilh) [reply swagger-backend.yaml:42]: that would be a bit confusing, as the `{namespace}` could be interepreted as being the `{device-id}` and vice versa. I'd prefer to either remove the namespace and put it as part of the payload or keep it as it is.

Comment (jordigilh) [reply swagger-backend.yaml:42]: I'll take `/finalize` :smile:

File: swagger-backend.yaml (hunk 2 above)
Comment (jordigilh) [reply swagger-backend.yaml:118]: Sounds good. Thanks for the suggestion.

File: swagger-backend.yaml (hunk 1 above)
Comment (jakub-dzon) [reply swagger-backend.yaml:42]: OK. That's fine.

**Comment by jordigilh** [comment]
> The heartbeat operation is missing.

Thanks, I added the `UpdateHearbeat` endpoint.

File: swagger-backend.yaml (hunk 1 above)
Comment (jordigilh) [reply swagger-backend.yaml]: s/POST/PUT/ I realized that semantically doesn't make sense to do a POST if you haven't done a PUT first. So PUT it is.

**Review by jakub-dzon — COMMENTED** [review]
The swagger design LGTM. 
Please get rid of all the errors.

File: Makefile [hunk 5]
```diff
@@ -275,6 +275,7 @@ k8s-client-gen: client-gen lister-gen informer-gen ## Generate typed client for
 
 validate-swagger: ## Validate swagger
 	$(DOCKER) run --rm -v $(PWD)/.spectral.yaml:/tmp/.spectral.yaml:z -v $(PWD)/swagger.yaml:/tmp/swagger.yaml:z stoplight/spectral lint --ruleset "/tmp/.spectral.yaml" /tmp/swagger.yaml
+	$(DOCKER) run --rm -v $(PWD)/.spectral.yaml:/tmp/.spectral.yaml:z -v $(PWD)/swagger-backend.yaml:/tmp/swagger-backend.yaml:z stoplight/spectral lint --ruleset "/tmp/.spectral.yaml" /tmp/swagger.yaml
```
Comment (masayag) [inline Makefile:278]: iinw, the path (the last argument) should be changed to `/tmp/swagger-backend.yaml` which is the path to which the yaml file is mounted in the image?

another suggestion is to run validation for both files within the same container. spectral usage should support several files as input.
```
spectral <command>

Commands:
  spectral lint [documents..]  lint JSON/YAML documents from files or URLs
```

File: swagger-backend.yaml [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    required:
+      - message
+    properties:
+      message:
+        type: string
+  device-configuration-response:
+    type: object
+    properties:
+      message:
+        type: string
```
Comment (masayag) [inline swagger-backend.yaml:26]: pls add description for the properties.
it will be easier to understand what each `*-response` returns

**Review by masayag — CHANGES_REQUESTED** [review]

File: swagger-backend.yaml (hunk 6 above)
Comment (jordigilh) [reply swagger-backend.yaml:26]: Understood

Comment (jordigilh) [reply swagger-backend.yaml:38]: Done.

File: Makefile (hunk 5 above)
Comment (jordigilh) [reply Makefile]: Done.

File: swagger-backend.yaml (hunk 6 above)
Comment (masayag) [reply swagger-backend.yaml:39]: Thank you. this makes the API very clear :+1:

**Review by jakub-dzon — APPROVED** [review]

**Pull request merged**